### PR TITLE
Consistently hairpin ICMP replies

### PR DESCRIPTION
**Pull request by wingo**
Refactor the lwaftr to consistently pass outgoing packets through a
common helper procedure which consistently checks if the packet needs
hairpinning, and consistently drops the packet to which the outgoing
packet is a reply.

Fixes:
- Memory leak for incoming IPv4 packets that don't match the binding
  table.
- Fix hairpinning of ICMP error messages for non-existent IPv4 hosts.
  Before, these errors always were going out the public IPv4 interface
  instead of being hairpinned if needed.
- Memory leak for incoming IPv6 packets that don't match the binding
  table.
- Memory leak for incoming IPv4 packets whose TTL is 0 or 1.
- Fix hairpinning of ICMP error messages for incoming IPv4 packets
  whose TTL is 0 or 1.  Before, these errors always were going out the
  public IPv4 interface instead of being hairpinned if needed.
- Fix hairpinning of ICMP error messages for packets with the DF bit
  that would cause fragmentation.  Likewise these were always going out
  the public interface.

### CODE REVIEW

File: src/apps/lwaftr/lwaftr.lua [hunk 1]
```diff
@@ -254,16 +254,48 @@ local function in_binding_table(lwstate, ipv6_src_ip, ipv6_dst_ip, ipv4_src_ip,
...
-   return transmit(lwstate.o4, icmp_dis)
+local function drop_ipv4_packet_to_unreachable_host(lwstate, pkt, to_ip)
+   if lwstate.policy_icmpv4_outgoing == lwconf.policies['DROP'] then
+      -- ICMP error messages off by policy; silently drop.
+      return drop(pkt)
+   end
+
+   if get_ipv4_proto(get_ethernet_payload(pkt)) == proto_icmp then
+      -- FIXME: We should send dst/host unreachable for ICMP
+      -- messages, unless the ICMP message is an error message.
```
Comment (wingo) [inline src/apps/lwaftr/lwaftr.lua]: Following the discussion in #263, I guess I should change this FIXME to refer to the RFC.

**Comment by wingo** [comment]
Needs tests for the new hairpinning cases.

File: src/apps/lwaftr/lwaftr.lua (hunk 1 above)
Comment (kbara) [inline src/apps/lwaftr/lwaftr.lua]: If it does, the RFC it refers to should be the one about not replying with ICMP to ICMP, rather than the lw4o6 RFC.

**Comment by kbara** [comment]
The code so far looks good; I'm looking forward to seeing the tests.

**Comment by wingo** [comment]
Rebased on top of lwaftr_starfruit to deal with conflicts from #265.  Otherwise the diff is the same.

**Comment by kbara** [comment]
Still LGTM; thanks for doing this.

**Comment by wingo** [comment]
This branch also changes softwire selection for incoming ICMPv4 ping from the internet: it always treats it as if the port were 0.  This will cause a drop if the IPv4 has reserved ports, and otherwise enables ping to B4's whose IPv4 addresses are not port-restricted.  Hairpinned ping seems to work too.

**Comment by wingo** [comment]
I still don't have tests for the error cases, or cases for vlans.  It's been mega-slow-going tho, excruciating really -- I am going to add one or two more tests and then I really have to move on to something else.

**Comment by wingo** [comment]
OK this one is totes ready to go, @kbara please review :)  I don't have exhaustive tests for all of the specific things I mention in the description but they all share the same codepath with the last commit that I added, so the code at least is tested.

**Comment by kbara** [comment]
Could you write a quick summary here of what is/isn't tested? :-)

**Comment by kbara** [comment]
This doesn't have equal test coverage for vlans, which is a blocker. It's super-easy to do, though - add the pcap names to diego's script to add vlans, generate the vlan packets, and nearly copy-paste these tests into end-to-end-vlan.sh (the variable names need slight tweaking, and it needs a vlan conf file). (Yes, this is inelegant.) It'll take maybe 15 minutes, if that.

**Comment by wingo** [comment]
Not tested:
 * The memory leak fixes.
 * Hairpinning of ICMP error messages for the following cases: non-existent IPv4 hosts, packets with the DF bit that would cause fragmentation, though see caveat 1 below.
 * Incoming ICMP ping to a non-port-restricted IPv4 address, with or without a corresponding softwire, though see caveat 2.

Tested:
 * Hairpinning of ICMP error messages for incoming IPv4 packets whose TTL is 0 or 1, in https://github.com/Igalia/snabbswitch/commit/e525074bc82627b36dcc02752fb064eab95ef911.  Caveat 1: This code path is the same as the above untested cases, just a different input.
 * Hairpinned ICMPv4 echo request and reply to non-port-restricted B4.  Caveat 2: Hairpinned ICMPv4 is a special case of incoming ICMPv4 from the internet so I think this code path is tested adequately as well.
 * Hairpinned ICMPv4 echo request and reply to an IPv4 that is managed by the lwaftr but for which there is no binding table entry. Caveat 2 also.

**Comment by kbara** [comment]
I've audited the tests. LGTM for merging, though I'd appreciate if you'd create an issue listing tests you think it should have but doesn't. Not having tests we should have isn't ideal, but forgetting about them entirely would be worse. :-)

**Comment by wingo** [comment]
Thanks for the review, and thanks again for helping with the vlan packets!

**Pull request merged**